braindecode/datautil/windowers.py

```diff
@@ -0,0 +1,356 @@
+            raw, mapping, chunk_duration=self.stride_samples / fs,
+        )
+        print(events[:,0])
+        if not self.drop_last_samples:
+            last_events = self._include_last_samples(raw, mapping)
+            print("last_events", last_events)
+            events = np.concatenate((events, last_events), axis=0)
+            # Reorder events
+            events = events[np.argsort(events[:, 0], axis=0)]
+
+        metadata = {
+            "event_onset_idx": events[:, 0],
+            "trial_number": range(len(events)),
+            "target": events[:, -1],
+        }
+        metadata = pd.DataFrame(metadata)
+        # metadata['subject'] =
+
+        return mne.Epochs(
+            raw,
+            events,
+            tmin=self.trial_start_offset_samples / fs,
+            tmax=(self.window_size_samples - 1) / fs,
+            baseline=None,
+            preload=False,
+            metadata=metadata,
+        )
+
+
+class Windower(object):
+    """
+    A windower that creates a mne Epochs objects. Therefore, it fits supercrops
+    of supercrop_size_samples in trial_start_offset_samples to
+    trial_stop_offset_samples separated by supercrop_stride_samples. If the last
+    supercrop does not end at trial_stop_offset_samples, creates another
+    overlapping supercrop that ends at trial_stop_offset_samples.
+
+    in mne: tmin (s)                    trial onset        onset + duration (s)
+    trial:  |--------------------------------|--------------------------------|
+    here:   trial_start_offset_samples                trial_stop_offset_samples
+
+    Parameters
+    ----------
+    trial_start_offset_samples: int
+        start offset from original trial onsets in samples
+    trial_stop_offset_samples: int
+        stop offset from original trial onsets in samples
+    supercrop_size_samples: int
+        supercrop size
+    supercrop_stride_samples: int
+        stride between supercrops
+    mapping: dict(str: int)
+        mapping from event description to target value
+    """
+    def __init__(self, trial_start_offset_samples, trial_stop_offset_samples,
+                 supercrop_size_samples, supercrop_stride_samples,
+                 drop_samples=False, mapping=None):
+        self.start_offset = trial_start_offset_samples
+        self.stop_offset = trial_stop_offset_samples
+        assert supercrop_size_samples > 0, (
+            "supercrop size has to be larger than 0")
+        self.size = supercrop_size_samples
+        assert supercrop_stride_samples > 0, (
+            "supercrop stride has to be larger than 0")
+        self.stride = supercrop_stride_samples
+        self.drop_samples = drop_samples
+        self.mapping = mapping
+
+    # TODO: handle case we don't get a raw
+    def __call__(self, raw):
+        # EventWindower: if annotations are given in raw, use them to get onsets
+        assert hasattr(raw, "annotations") and hasattr(raw.annotations, "onset")
```

Comment:
I think every mne.Raw object has an annotations attribute with an onset attribute. This statement will thus always be true. If annotations is not set, raw.annotations.onset will be a list of length 0. This will probably lead to a later error, e.g. when adding the `stop_offset` in `_supercrop_starts` to an empty list, resulting in another empty list.

If annotations are mandatory, you could go for `assert len(raw.annotations.onset) != 0`.